### PR TITLE
[GSoC] Refactor preferences initialization

### PR DESCRIPTION
**Pull request by BrayanDSO**
~~On top of #11866. I should rebase once that is merged.~~

## Purpose / Description

Preferences initialization was a bit messy. Every screen ran [initAllPreferences](https://github.com/ankidroid/Anki-Android/blob/abd1f8bfc316cdf11aba51af7bda7f5b05c493d7/AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt#L205) and [initPreference](https://github.com/ankidroid/Anki-Android/blob/abd1f8bfc316cdf11aba51af7bda7f5b05c493d7/AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt#L225) multiple times, making things quite hard to follow. 

To keep things tidy, concise, and keep the preferences' configuration responsability on themselves and on their screens, I've done some refactors on the preferences activity

## Approach

1. Extract the preferences initialization to their screens on the `Init X preference individually` commits
- The required changes that escape the previous 2 patterns are marked on the commits as a `**` on the message beginning and are described as well

## How Has This Been Tested?

Manually opened/changed all preferences and it still works correctly. I should upload a video of it later.

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by BrayanDSO** [comment]
Split into #11873

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 1]
```diff
@@ -608,6 +603,15 @@ class Preferences : AnkiActivity() {
             }
             // Build languages
             initializeLanguageDialog(screen)
+
+            // Paste PNG
+            requirePreference<SwitchPreference>(R.string.paste_png_key).apply {
+                isChecked = col.get_config("pastePNG", false)!!
+                setOnPreferenceChangeListener { _, newValue ->
+                    col.set_config("pastePND", newValue)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:506]: typo: `pastePND` -> `pastePNG`

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 2]
```diff
@@ -475,11 +474,6 @@ class Preferences : AnkiActivity() {
                     DAY_OFFSET -> {
                         preferencesActivity.setDayOffset((pref as SeekBarPreferenceCompat).value)
                     }
-                    PASTE_PNG -> {
-
-                        preferencesActivity.col.set_config("pastePNG", (pref as SwitchPreference).isChecked)
-                        preferencesActivity.col.setMod()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:489]: Agreed - setMod is no longer necessary

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 3]
```diff
@@ -853,6 +849,16 @@ class Preferences : AnkiActivity() {
                 true
             }
             initializeCustomFontsDialog()
+
+            // Show estimate time
+            requirePreference<SwitchPreference>(R.string.show_estimates_preference).apply {
+                isChecked = col.get_config_boolean("estTimes")
+                setOnPreferenceChangeListener { _, newValue ->
+                    col.set_config("estTimes", newValue)
+                    col.setMod()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:797]: no need for `setMod`. This is now handled in `set_config`

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 4]
```diff
@@ -859,6 +854,15 @@ class Preferences : AnkiActivity() {
                     true
                 }
             }
+            // Show progress
+            requirePreference<SwitchPreference>(R.string.show_progress_preference).apply {
+                isChecked = col.get_config_boolean("dueCounts")
+                setOnPreferenceChangeListener { _, newValue ->
+                    col.set_config("dueCounts", newValue)
+                    col.setMod()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:806]: ditto - no need

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 5]
```diff
@@ -639,8 +634,15 @@ class Preferences : AnkiActivity() {
             val col = col!!
 
             // Learn ahead limit
-            requirePreference<NumberRangePreferenceCompat>(R.string.learn_cutoff_preference)
-                .setFormattedSummary(R.string.pref_summary_minutes)
+            requirePreference<NumberRangePreferenceCompat>(R.string.learn_cutoff_preference).apply {
+                setValue(col.get_config_int("collapseTime") / 60)
+                setFormattedSummary(R.string.pref_summary_minutes)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:559]: optional follow up: we should probably turn this into a set-only property

**Review by david-allison — CHANGES_REQUESTED** [review]
Refactoring:

I would like `setOnPreferenceChangeListener` to be simplified: 

* mot of these do not use the first parameter
* most of these only return `true`

----

remove most of the `setMod` calls, check `set_config()` and these are now superfluous

**Comment by BrayanDSO** [comment]
> You removed a lot of documentation, could this be moved to preferences.xml

Forgot to add those back. Added on the last commit.

> I would like setOnPreferenceChangeListener to be simplified via an extension method/property if possible

Really liked the idea. I'll follow it up after #11874

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 6]
```diff
@@ -684,6 +659,27 @@ class Preferences : AnkiActivity() {
             // Time to show question
             requirePreference<SeekBarPreferenceCompat>(R.string.timeout_question_seconds_preference)
                 .setFormattedSummary(R.string.pref_summary_seconds)
+
+            // New timezone handling
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:618]: My bad. Fixed it. I think the logic is correct but I should add a test for it on this PR when I get the time.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 5 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:567]: If it is `setFormattedSummary(R.string.pref_summary_minutes)`, it is already set only.

If it is `NumberRangePreferenceCompat.setValue` I don't see why.

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt [hunk 7]
```diff
@@ -35,5 +44,35 @@ class NotificationsSettingsFragment : SpecificSettingsFragment() {
...
+                    )
+                    val alarmManager = requireActivity().getSystemService(ALARM_SERVICE) as AlarmManager
+                    alarmManager.cancel(intent)
+                }
+                true
+            }
+        }
+    }
+
+    private fun updateNotificationPreference(listPreference: ListPreference) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:67]: While you're moving this function, do you think you could add a javadoc to it. 
I suspect I can understand what it does, but I'd love to be certain I understood correctly, and that may help its next reader.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 8]
```diff
@@ -491,6 +491,9 @@ class Preferences : AnkiActivity() {
             initializeLanguageDialog(screen)
 
             // Deck for new cards
+            // Represents in the collections pref "addToCur": i.e.
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:494]: Regarding a25606021813aebe5f8281dce20ae511e55b939f, if it's not too much work, I'd appreciate if you could move the comment in teh same commit that removed the comment initially. It would be easier to follow the history of this text if it's even useful, and avoid an extra commit that's not really justified as far as history goes. Also, it'll clearly help to review that the right comment went to the right place

**Review by Arthur-Milchior — APPROVED** [review]
A lot of small nit, preference to change commit messages. Nothing blocking, so I approve. If you accept to do some of those changes, I'm fine to merge after (using only a simple diff between current HEAD and the future one to check changes)

I'm surprised there is no way to create a simple function that would avoid repeating code in all of the new listener, but there is so many little variation that I admit that writing all of them explicitly is better.

I don't know if you have an opinion, but I'm not a fan of a 1358 lines long file with that many subclasses. I believe code would be easier to read, navigate and potentially edit if it were split into one file by subpages of the preferences. As far as I understand, with your change it becomes even easier. 
Also, if I ever want to look at the code of a specific preference, it'll be easier to open the correct file than to navigate and find the right class.
If we go this way, I'd suggest creating a subdirectory  in /anki named /preferences, containing the top level preferences and each subpreferences (and probably SpecificSettingsFragment and SettingsFragment)

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 5 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:567]: It's a `fun`, not a `var`:

https://github.com/ankidroid/Anki-Android/blob/8c251f9810b929e38190a7b47d6e60773729ed70/AnkiDroid/src/main/java/com/ichi2/preferences/SeekBarPreferenceCompat.kt#L102-L104

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:567]: EDIT: Apparently Kotlin doesn't support this. My bad

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt (hunk 7 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:68]: I expect this to be gone when @prateek-singh-3212 does the UI for the new notifications screen, so I'd like to avoid doing the temporary work of adding a comment to it.

File: AnkiDroid/src/main/java/com/ichi2/preferences/NumberRangePreferenceCompat.kt [hunk 9]
```diff
@@ -205,14 +205,6 @@ open class NumberRangePreferenceCompat : EditTextPreference {
             editText.filters = arrayOf(*editText.filters, InputFilter.LengthFilter(numberRangePreference.maxDigits))
         }
 
-        override fun onDialogClosed(positiveResult: Boolean) {
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/preferences/NumberRangePreferenceCompat.kt:208]: Yes, I've looked git blame and found no reason to keep it. Explained better on the commit message:
```
The only difference between NumberRangePreferenceCompat.onDialogClosed and super.onDialogClosed is that NumberRangePreferenceCompat's method doesn't call the preference change listener.

As some of the current preferences need a preference change listener and there isn't a reason to ignore it (nor it had one when the method was introduced), I'm removing the override.
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 10]
```diff
@@ -222,22 +222,7 @@ class Preferences : AnkiActivity() {
     }
 
     private fun initPreference(pref: Preference) {
-        // Load stored values from Preferences which are stored in the Collection
-        if (sCollectionPreferences.contains(pref.key)) {
-            val col = col
-            if (col != null) {
-                try {
-                    when (pref.key) {
-                        AUTOMATIC_ANSWER_ACTION -> (pref as ListPreference).setValueIndex(col.get_config(AutomaticAnswerAction.CONFIG_KEY, 0.toInt())!!)
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:236]: > If it can't be done because one can't compile an empty when, or an empty try

this is the case. I tend to prefer removing it altogether and explaining why on the commit message than adding useless changes just for commit splitting.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 2 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:489]: Done

File: AnkiDroid/src/main/java/com/ichi2/preferences/NumberRangePreferenceCompat.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/preferences/NumberRangePreferenceCompat.kt:208]: Base method handles the same functionality as this method - seems fine to me.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 11]
```diff
@@ -222,22 +222,7 @@ class Preferences : AnkiActivity() {
     }
 
     private fun initPreference(pref: Preference) {
-        // Load stored values from Preferences which are stored in the Collection
-        if (sCollectionPreferences.contains(pref.key)) {
-            val col = col
-            if (col != null) {
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: We need to handle the collection == null case in the preferences overview, and in Advanced (where the AnkiDroid directory is). The advanced preferences window can be opened in the case of collection corruption, or being unable to find the folder which contains the collection.

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 12]
```diff
@@ -1380,13 +1387,6 @@ class Preferences : AnkiActivity() {
          */
         private const val DAY_OFFSET = "dayOffset"
 
-        /**
-         * Represents in Android preference the collection's configuration "pastePNG" , i.e.
-         * whether to convert clipboard uri to png format or not.
-         * TODO: convert to png if a image file has transparency, or at least if it supports it.
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:1393]: Can we have a GH issue for this instead? I think it would be more easily discoverable this way.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: Does anyone have a corrupted collection for me to reproduce this?

And I'd like to know exactly what you meant by `preferences overview`. Is that the preferences first pages with the top headers (General, Sync, Reviewing, etc)? Because it's what I initially think when I see "overview".

[the remark this answers is not in the document]
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: I don't have a corrupted collection but if you want to make a test you could check for 'illFatedDb' in the tests directory and see how we corrupt one programmatically - combining the existing test collections we have as fixtures and the same idea of opening the file and scribbling in it (that is: overwriting a chunk with :hankey: data) might be useful? That test was easier for me to write then I thought when I tried it so I'm not trying to suggest something really painful, it might be easy to do

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:1393]: Sure, bring it it and mark with `Keep Open`

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: After spending weeks thinking about it, I've finally designed a reliable process to create a corrupted collection. Beware that it may me too complex.

Step 1: rename any file to `collection.anki2`
Step 2: there isn't a step 2

This is the default patterns we can follow with a corrupt collection:
https://user-images.githubusercontent.com/69634269/179859192-daa6f198-bdcd-4355-8578-de3b0fb9658c.mp4

So we can't go to the preferences with a corrupt collection.

But you are still right that there is a way to go to the advanced screen on a startup error

https://github.com/ankidroid/Anki-Android/blob/31cfbc914a7467eaeeeb74a461821f1166d707f5/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt#L478-L483

And DIRECTORY_NOT_ACCESSIBLE is thrown here

https://github.com/ankidroid/Anki-Android/blob/1373897c73a613d745958ad0fdbc9d32cfdae4a8/AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt#L45-L46

Which will depends on isCurrentAnkiDroidDirAccessible returning false

https://github.com/ankidroid/Anki-Android/blob/5a586f622a06507e9a8aeae92c49b8058fd76cc9/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java#L327-L335

Which is only the case if initializeAnkiDirectory fails

https://github.com/ankidroid/Anki-Android/blob/5a586f622a06507e9a8aeae92c49b8058fd76cc9/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java#L302-L320

If there write permission hasn't been granted, it doesn't lead to the advanced screen. 

So the only other way is to fail `dir.mkdirs` and I don't know how to fail it on a real use scenario, so I need help here to conclude if it's possible or not to get to the advanced screen with a unnacessible directory.

If it's not possible, there isn't a issue to solve here. If it is possible, I think that opening Android's directory picker is a better UX than opening the advanced screen.

[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: Kindly related, I've discovered another issue while researching this:

DatabaseErrorDialog shouldn't be cancellable by outside touches. 

https://user-images.githubusercontent.com/69634269/179863188-2bd71437-4189-468f-a0f5-8159e24cace9.mp4

It leads to crashes if I open the Card Browser, Statistics and a lot of other things.

Task here is to add `cancelOnTouchOutside(false)` to `DatabaseErrorDialog`

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:1393]: Shouldn't this have solved the TODO? https://github.com/ankidroid/Anki-Android/pull/9213

[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:1393]: Totally right, sorry for the misdirection

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:228]: Okay, now that I look better at it, I actually haven't added a `col!!` call on `AdvancedSettingsFragment` + you can't navigate back to the preferences main screen as the fragment is called with `getSubscreenIntent`, so I really can't see the issue here. 

There was another `col!!` call on `AdvancedSettingsFragment`, but I've fixed it on the last commit just in case

**Comment by david-allison** [comment]
@Arthur-Milchior @mikehardy Treat this as an 'approve' if the collection issue was fixed.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 13]
```diff
@@ -932,7 +932,7 @@ class Preferences : AnkiActivity() {
             }
             setupContextMenuPreference(CardBrowserContextMenu.CARD_BROWSER_CONTEXT_MENU_PREF_KEY, R.string.card_browser_context_menu)
             setupContextMenuPreference(AnkiCardContextMenu.ANKI_CARD_CONTEXT_MENU_PREF_KEY, R.string.context_menu_anki_card_label)
-            if (col!!.schedVer() == 1) {
+            if (col != null && col!!.schedVer() == 1) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:935]: This means that the previous commits may lead to a crash, and we know it. 
As a matter of principle, I would have preferred this commit to arrive sooner to avoid it.
In practice, since it does not break tests, it won't impact bisect, and it's not like we'll sheep intermediary version, so no practical reason to block here.

**Review by Arthur-Milchior — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 14]
```diff
@@ -684,6 +659,24 @@ class Preferences : AnkiActivity() {
             // Time to show question
             requirePreference<SeekBarPreferenceCompat>(R.string.timeout_question_seconds_preference)
                 .setFormattedSummary(R.string.pref_summary_seconds)
+
+            // New timezone handling
+            requirePreference<SwitchPreference>(R.string.new_timezone_handling_preference).apply {
+                isChecked = col.sched._new_timezone_enabled()
+                isEnabled = col.schedVer() > 1
+                setOnPreferenceChangeListener { _, newValue ->
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:622]: #11874 is built on top of this and I would like to avoid the chance of getting conflicts and having to spend time fixing them.
